# Analyses of haploblocks 

* Arg_chr6_pcaplot.R
   
   Plot of arg06.01/02 region PCA
* GEA_inv_snp_compared.R
   
   Comparison of haploblocks and SNPs GEA, looking for enrichment of associations in haploblocks
* GWAS_inv_snp_compared.R

   Comparison of haploblocks and SNPs GWA, looking for enrichment of associations in haploblocks
* argophyllus.chr6.dtf.R
   
   Plot of arg06.01 (10 MB region) and flowering time.
* argophyllus.chr6.missingdata.R

   Plot of missing data in arg06.01 region
 
*  argophyllus_ftcoverage.R

   Plot of sequence coverage in FT genes in arg06.01
* gea_gwas_enrichment_figure.R

   Plot of comparison of GEA and GWA enrichment for associations in haploblocks
* genetic_maps.R

   Plotting and comparison of H. annuus genetic maps.
* gwas_gea_pleiotropy_plot.R

   Figure of GWA/GEA associations by category.
* inversion.pcasites.genotypes.R

   Plotting diagnostic markers for haploblocks to see potential recombination.
* inversion_age.R
   
   Plotting haploblock divergence age estimates.
* inversion_locations.R

   Plotting haploblock locations
* petiolaris.inversion.synchronization.R

   Synchronizing haploblocks identified in different subsets of petiolaris samples (petpet, petfal and both together).
* run_ld_windows_*.sh
   
   Measuring LD across the genome for each sub/species 
* test_haploblock_ld_annuus.sh

   Measuring LD for the major haploblock genotype, across the haploblock chromosome. 

*  visualize_ha412_inversions_forfigure_v2.R

   Plotting MDS, PCA, heterozygosity, and LD for each haploblock
